include/merlin/core_kernels/lookup_kernels.cuh

```diff
@@ -0,0 +1,722 @@
+              sm_possible_keys[SAME_BUF(i)] + (groupID * RESERVE + group_base),
+              key_ptr + key_pos, sizeof(K));
+          group_base += 1;
+        } else {
+          break;
+        }
+      } while (true);
+    } else {
+      K target_key = sm_target_keys[key_idx_block];
+      sm_counts[key_idx_block] = 0;
+      int found_vote = 0;
+      bool found = false;
+      do {
+        int digest_idx = __ffs(find_result) - 1;
+        if (digest_idx >= 0) {
+          find_result &= (find_result - 1);
+          int key_pos = rank * 4 + digest_idx;
+          K possible_key = key_ptr[key_pos];
+          if (possible_key == target_key) {
+            found = true;
+            sm_counts[key_idx_block] = 1;
+            sm_possible_pos[SAME_BUF(i)][groupID * RESERVE] = key_pos;
+            sm_possible_keys[SAME_BUF(i)][groupID * RESERVE] = possible_key;
+          }
+        }
+        found_vote = g.ballot(found);
+        if (found_vote) {
+          break;
+        }
+        found_vote = digest_idx >= 0;
+      } while (g.any(found_vote));
+    }
+    __pipeline_commit();
+
+    /* Step3: check possible keys, and prefecth the value and score */
+    if (i > 0) {
+      key_idx_block -= 1;
+      K target_key = sm_target_keys[key_idx_block];
+      int possible_num = sm_counts[key_idx_block];
+      sm_founds[key_idx_block] = 0;
+      S* score_ptr = CopyScore::get_base_ptr(sm_keys_ptr, key_idx_block);
+      V* value_ptr = sm_values_ptr[key_idx_block];
+      __pipeline_wait_prior(3);
+      int key_pos;
+      bool found_flag = false;
+      if (rank < possible_num) {
+        K possible_key =
+            sm_possible_keys[DIFF_BUF(i)][groupID * RESERVE + rank];
+        key_pos = sm_possible_pos[DIFF_BUF(i)][groupID * RESERVE + rank];
+        if (possible_key == target_key) {
+          found_flag = true;
+          CopyScore::ldg_sts(sm_target_scores + key_idx_block,
+                             score_ptr + key_pos);
+        }
+      }
+      int found_vote = g.ballot(found_flag);
+      if (found_vote) {
+        V* v_dst = sm_vector[DIFF_BUF(i)][groupID];
+        sm_founds[key_idx_block] = 1;
+        int src_lane = __ffs(found_vote) - 1;
+        int target_pos = g.shfl(key_pos, src_lane);
+        V* v_src = value_ptr + target_pos * dim;
+        CopyValue::ldg_sts(rank, v_dst, v_src, dim);
+      }
+    }
+    __pipeline_commit();
+
+    /* Step4: write back value and score */
+    if (i > 1) {
+      key_idx_block -= 1;
+      int key_idx_grid = blockIdx.x * blockDim.x + key_idx_block;
+      V* v_src = sm_vector[SAME_BUF(i)][groupID];
+      V* v_dst = values + key_idx_grid * dim;
+      int found_flag = sm_founds[key_idx_block];
+      __pipeline_wait_prior(3);
+      if (found_flag > 0) {
+        S score_ = CopyScore::lgs(sm_target_scores + key_idx_block);
+        CopyValue::lds_stg(rank, v_dst, v_src, dim);
+        founds[key_idx_grid] = true;
+        CopyScore::stg(scores + key_idx_grid, score_);
```

Comment:
Because the result of score and found information is stored in shared memory, to write them to the global memory in the end of the kernel, for coalesing memory access, targeting reduce memory traffic.